lib/spack/spack/solver/asp.py

```diff
@@ -889,7 +927,8 @@ def store(self, problem: str, result: Result, statistics: List, test: bool = Fal
                 tty.debug(f"Cache entry {cache_path} exists, will not be overwritten")
                 return
             cache_dict = {"results": result.to_dict(test=test), "statistics": statistics}
-            bytes_written = new.write(json.dumps(cache_dict))
+            new_bytes = open(new.name, "wb")
+            bytes_written = new_bytes.write(self._zip_cache_entry(json.dumps(cache_dict)))
```

Comment:
I think it's incorrect to open the same file a second time.

Reply:
we have `llnl.util.filesystem.write_tmp_and_move()` and a few others

Reply:
The idea here is to use the `FileCache` so I can drop in a Storage object with a similar API without having to re-write too much.

Reply:
Shouldn't be hard to add read/write_transaction(..., mode).

I don't really see the advantage of FileCache since it always copies the file that you're planning to update in-place. What's the reason you want to use FileCache?

Reply:
hm okay, filecache is mostly for locking, that's not very relevant for remotes

Reply:
I am working on a follow up that is a drop in replacement for the `FileCache` API. It is meant primarily to make adding remote caching easier later.

The web_util APIs are rather more limited and relies on global variables and is heavily tied to mirrors. Rather than a more significant refactor of that, I am working on a more generic interface for dealing with storage devices.

Reply:
`FileCache` is just a placeholder for now. Writing something else the is more configurable/better suited for this is probably not a bad idea.